### ISSUE TITLE
Discussion: Scaling Federated Clusters through multiplexing

### ISSUE THREAD

**Issue by hansonkd** [nats-io/nats-server#669]
I am working on a project that has different small "groups" of users. The cluster they connect to should be isolated from other groups. Right now this is accomplished by running multiple servers binding on different ports, but this does not scale.

What would be desirable is that you can run multiple servers on one machine and only listen on 2 ports (all client connections and the cluster connection). 

## Proposal:
0. Clusters have IDs
1. Add a `CLUSTER <Cluster-id>` message to the protocol. This message would be the first message sent by clients.
    a. When server accepts client connection, it checks for cluster message, and then routes the connection to the appropriate server.
2. Cluster connections are multiplexed using something like https://github.com/zenhotels/astranet

## Diagram

The colored boxes are `gnatsd.Server` instances that are dedicated to a certain cluster. Each `Server-Group` is one machine (meaning multiple servers are sharing the same socket). The white lines are mutiplexed connections for `INFO` chatter of the cluster.

<img width="800" alt="screen shot 2018-05-04 at 11 18 36 am" src="https://user-images.githubusercontent.com/496914/39632934-e9a5ec18-4f8c-11e8-8ef9-58919c30b664.png">

## Benefits:
1. Chatter about interests do not need to go to every node.
2. Can scale certain subgroups up or down.
3. Simplify devops for serving multiple clusters on few resources.

## Downsides:
1. Protocol change
2. Multiplexing overhead

## Questions
1. 10 servers of 1,000 connections on one machine about as efficient as running 10,000 connections on 1 server?

Best of all this change can be done without affecting the current gnatsd operation. The only thing that it requires from gnatsd is the ability to specify listeners for both client and router.

Basically when starting a server, do something like:

```go
// ClientListener would simply be wrapper around a channel that you can write net.Conn into
clientListener, _ := NewClientListener(groupName)
routeListener, _ := astraNet.Bind("", groupName))

s := server.New(opts)
s.SetClientListener(clientListener)
s.SetRouteListener(routeListener)
s.SetRouteDialer(astraNetDialer)
...
```

## Realistic Proposal
You dont to accept this whole proposal to help unblock. Instead of making this a Protocol change, just allow us to pass in our own listeners when we start the server  (see https://github.com/nats-io/gnatsd/issues/668). Just having `SetClientListener` and `SetRouteListener` would be enough to write external libraries to extend the gnatsd server to do this.

**Comment by antmanler** [comment]
we had similar use case, instead modifying protocol we extract parser from gnatsd and implemented a gateway to route different connections based on the info when client sends ‘CONNECT’ command.
project https://github.com/antmanler/gnatsd-gw

**Comment by hansonkd** [comment]
I thought I would share my progress: Multiplexed gnatsd is working! Currently it uses fixed 32character cluster names.

It uses this pull request: https://github.com/nats-io/gnatsd/pull/670

Once gnatsd allows us to control the ports we can do interesting things.

On the server side:


Custom Dialers and Listeners
```go

// When dialing to a cluster, check if a multiplex stream is open to that address, if it is reuse it. Then send the `CLUSTER <>` message
type RoutedClusterDialer struct {
    d *SessionDialer
    g string
}
func (d * RoutedClusterDialer) Dial(network, address string) (net.Conn, error) {
    session, err := d.d.DialSession(network, address)
    if err != nil {
        return nil, err
    }
    conn, err := session.Open()
    if err != nil {
        return nil, err
    }
    msg := []byte(fmt.Sprintf("CLUSTER %s", d.g))
    conn.Write(append(msg, crlfBytes...))
    return conn, nil
}

// Caches 
type SessionDialer struct {
    d proxy.Dialer
    sessions map[string]*yamux.Session
    sync.Mutex
}
func (d *SessionDialer) DialSession(network, address string) (*yamux.Session, error) {
    d.Lock()
    defer d.Unlock()
    if session, ok := d.sessions[address]; ok {
        return session, nil
    }
    conn, err := d.d.Dial(network, address)
    if err != nil {
        return nil, err
    }
    session, err := yamux.Client(conn, nil)
    if err != nil {
        return nil, err
    }
    d.sessions[address] = session
    return session, nil
}

// A simple listener that listens for `net.Conn` on a Chan in `Accept`
type ChannelListener struct {
    addr net.Addr
    connChan chan net.Conn
}
func (l ChannelListener) Close() error {
    close(l.connChan)
    return nil
}
func (l ChannelListener) Accept() (net.Conn, error) {
    conn, ok := <-l.connChan
    if ok {
       return conn, nil
    }
    return nil, errors.New("Connection closed")
}
func (l ChannelListener) Addr() net.Addr {
    return l.addr
}
```


Now the Supervisor

```go
type ServerSupervisor struct {
    clientListener net.Listener
    clusterListener net.Listener
    clientRoutes map[string](chan net.Conn)
    clusterRoutes map[string](chan net.Conn)
    servers map[string](*server.Server)
    sessionDialer *SessionDialer
    serverOpts *server.Options
    sync.Mutex
}
```

Supervisor Methods
```go
func NewServerSupervisor(sup *core.Supervisor, opts *server.Options, clientListener net.Listener, clusterListener net.Listener, routeDialer proxy.Dialer) *ServerSupervisor {
    return &ServerSupervisor{
        supervisor: sup,
        clientListener: clientListener,
        clusterListener: clusterListener,
        clientRoutes: make(map[string](chan net.Conn), 0),
        clusterRoutes: make(map[string](chan net.Conn), 0),
        servers: make(map[string](*server.Server), 0),
        serverOpts: opts,
        sessionDialer: &SessionDialer{d: routeDialer, sessions: make(map[string]*yamux.Session)},
    }
}
func (sup *ServerSupervisor) Start() {
    fmt.Println("Start")
    go sup.ListenAndRoute(sup.clientListener, sup.clientRoutes)
    go sup.ListenAndRouteMultiplex(sup.clusterListener, sup.clusterRoutes)
    sup.supervisor.Start()
}

// The first connection on multiplex opens a new session. You need pass that listener to the accept loop.
func (sup *ServerSupervisor) ListenAndRouteMultiplex(l net.Listener, routes map[string](chan net.Conn)) error {
    for {
        conn, err := l.Accept()
        if err != nil {
            return err
        }
        sessionListener, err := yamux.Server(conn, nil)
        if err != nil {
            return err
        }
        go sup.ListenAndRoute(sessionListener, routes)
    }
}
func (sup *ServerSupervisor) ListenAndRoute(l net.Listener, routes map[string](chan net.Conn)) error {
    buf := make([]byte, clusterMsgSize)
    for {
        conn, err := l.Accept()
        if err != nil {
            return err
        }
        conn.SetReadDeadline(time.Now().Add(timeoutDuration))
        _, err = conn.Read(buf)
        if err != nil {
            conn.Close()
            continue
        }
        conn.SetReadDeadline(time.Time{})

        if !bytes.HasPrefix(buf, []byte("CLUSTER")) {
            conn.Close()
            continue
        }
        g := buf[clusterMsgSize-32-2:clusterMsgSize-2]

        sup.Lock()
        r, ok := routes[string(g)]
        sup.Unlock()
        if ok {
            r <- conn
        } else {
            conn.Close()
            continue
        }
    }
}
```

Finally add your servers
```
func (sup *ServerSupervisor) AddGroup(g string, routeStr string) error {
    _, ok := sup.clientRoutes[g]
    if !ok {
        clientChan := make(chan net.Conn, 64)
        routeChan := make(chan net.Conn, 64)
        opts := sup.serverOpts.Clone()
        if routeStr != "" {
            opts.Routes = server.RoutesFromStr(routeStr)
        }
        opts.ClientListener = &ChannelListener{addr: sup.clientListener.Addr(), connChan: clientChan}
        opts.RouteListener = &ChannelListener{addr: sup.clusterListener.Addr(), connChan: routeChan}
        opts.RouteDialer = &RoutedDialer{d: sup.sessionDialer, g: g}
        serv := server.New(opts)
        serv.ConfigureLogger()
        go serv.Start()

        sup.Lock()
        sup.clientRoutes[g] = clientChan
        sup.clusterRoutes[g] = routeChan
        sup.servers[g] = serv
        sup.Unlock()
    }
    return nil
}
```

@nats-io/core

**Comment by derekcollison** [comment]
We will be adding account support to NATS in the near future which will provide the isolation you are seeking. This will allow for secure and protected multi-tenant use on a single server/cluster.

**Comment by bruth** [comment]
Accounts were introduced in NATS 2.0. Currently at 2.9.17 😄
